// react-vr-web/js/Modules/RCTAudioModule.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

/**
 * @class RCTAudioModule
 * @extends Module
 */

import {VRAudioContext, VRAudioComponent} from 'ovr-audio';
import Module from './Module';

export default class RCTAudioModule extends Module {
  constructor(rnctx) {
    super('RCTAudioModule');
    this.audioContext = new VRAudioContext();
    this._audioDefs = {};
    this._components = {};
    this._rnctx = rnctx;
  }

  /**
   * Add a new handle to the audio module
   * @param {string} handle - The audio handle.
   */
  addHandle(handle, audioConfig) {
    const component = new VRAudioComponent(this.audioContext, audioConfig);
    this._components[handle] = component;
    this._audioDefs[handle] = this._createAudioDef();
    component.onMediaReady = this._onMediaReady.bind(this, handle);
    component.onMediaEnded = this._onMediaEnded.bind(this, handle);
  }

  _createAudioDef() {
    return {
      streamingType: 'buffer',
    };
  }

  _onMediaReady(handle) {
    // Emit audio ready event to react
    this._rnctx.callFunction(
      'RCTDeviceEventEmitter',
      'emit',
      ['onAudioReady', handle]);
  }

  _onMediaEnded(handle) {
    // Emit audio ended event to react
    this._rnctx.callFunction(
      'RCTDeviceEventEmitter',
      'emit',
      ['onAudioEnded', handle]);
  }

  /**
   * Set the audio url
   * @param {string} handle - The audio handle.
   */
  setUrl(handle, url) {
    this._audioDefs[handle].src = url;
  }

  /**
   * load the audio
   * @param {string} handle - The audio handle.
   */
  load(handle) {
    this._components[handle].setAudio(this._audioDefs[handle]);
  }

  /**
   * play the video
   * @param {string} handle - The video handle.
   */
  play(handle) {
    this._components[handle].play();
  }

  /**
   * stop the video
   * @param {string} handle - The video handle.
   */
  stop(handle) {
    this._components[handle].stop();
  }

  /**
   * dispose the video
   * @param {string} handle - The video handle.
   */
  dispose(handle) {
    this._components[handle].dispose();
    delete this._components[handle];
    delete this._audioDefs[handle];
  }

  frame(camera) {
    this.audioContext.frame(camera);
  }
}
