// ovrui/lib/OVRUI.js
/*
 * Copyright (c) 2016-present, Oculus, LLC
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Player = exports.Overlay = exports.CSSBorderGeometry = exports.MergedBufferGeometry = exports.FourByFourRectGeometry = exports.UIView = exports.UIViewEvent = exports.UIViewEventType = exports.GuiSysHitType = exports.GuiSysEvent = exports.GuiSysEventType = exports.GuiSys = exports.SDFFONT_MARKER_COLOR = exports.TOP = exports.RIGHT_LINE = exports.RIGHT = exports.LEFT = exports.CENTER_LINE = exports.CENTER_FIXEDHEIGHT = exports.CENTER = exports.BOTTOM = exports.BASELINE = exports.wrapLines = exports.measureText = exports.loadFont = exports.BitmapFontGeometry = exports.VREffect = exports.VRControls = exports.DeviceOrientationControls = exports.AppControls = undefined;

var _AppControls = require('./Control/AppControls');

var _AppControls2 = _interopRequireDefault(_AppControls);

var _DeviceOrientationControls = require('./Control/DeviceOrientationControls');

var _DeviceOrientationControls2 = _interopRequireDefault(_DeviceOrientationControls);

var _VRControls = require('./Control/VRControls');

var _VRControls2 = _interopRequireDefault(_VRControls);

var _VREffect = require('./Control/VREffect');

var _VREffect2 = _interopRequireDefault(_VREffect);

var _FourByFourRect = require('./FourByFourRect/FourByFourRect');

var _MergedBufferGeometry = require('./FourByFourRect/MergedBufferGeometry');

var _CSSBorderGeometry = require('./FourByFourRect/CSSBorderGeometry');

var _SDFFont = require('./SDFFont/SDFFont');

var _GuiSys = require('./UIView/GuiSys');

var _GuiSys2 = _interopRequireDefault(_GuiSys);

var _GuiSysEvent = require('./UIView/GuiSysEvent');

var _UIView = require('./UIView/UIView');

var _UIView2 = _interopRequireDefault(_UIView);

var _Player = require('./Player/Player');

var _Player2 = _interopRequireDefault(_Player);

var _Overlay = require('./Player/Overlay');

var _Overlay2 = _interopRequireDefault(_Overlay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Controls
exports.AppControls = _AppControls2.default; /*
                                              * Copyright (c) 2016-present, Oculus, LLC
                                              * All rights reserved.
                                              *
                                              * This source code is licensed under the license found in the LICENSE file in
                                              * the root directory of this source tree.
                                              */

// This file will eventually become the entry point of the "ovrui" npm module
// It acts as a single point of entry for the various exports

exports.DeviceOrientationControls = _DeviceOrientationControls2.default;
exports.VRControls = _VRControls2.default;

// VR camera effect

exports.VREffect = _VREffect2.default;

// SDF Fonts

exports.BitmapFontGeometry = _SDFFont.BitmapFontGeometry;
exports.loadFont = _SDFFont.loadFont;
exports.measureText = _SDFFont.measureText;
exports.wrapLines = _SDFFont.wrapLines;
exports.BASELINE = _SDFFont.BASELINE;
exports.BOTTOM = _SDFFont.BOTTOM;
exports.CENTER = _SDFFont.CENTER;
exports.CENTER_FIXEDHEIGHT = _SDFFont.CENTER_FIXEDHEIGHT;
exports.CENTER_LINE = _SDFFont.CENTER_LINE;
exports.LEFT = _SDFFont.LEFT;
exports.RIGHT = _SDFFont.RIGHT;
exports.RIGHT_LINE = _SDFFont.RIGHT_LINE;
exports.TOP = _SDFFont.TOP;
exports.SDFFONT_MARKER_COLOR = _SDFFont.SDFFONT_MARKER_COLOR;

// UIView

exports.GuiSys = _GuiSys2.default;
exports.GuiSysEventType = _GuiSysEvent.GuiSysEventType;
exports.GuiSysEvent = _GuiSysEvent.GuiSysEvent;
exports.GuiSysHitType = _GuiSysEvent.GuiSysHitType;
exports.UIViewEventType = _GuiSysEvent.UIViewEventType;
exports.UIViewEvent = _GuiSysEvent.UIViewEvent;
exports.UIView = _UIView2.default;

// FourByFourRectGeometry

exports.FourByFourRectGeometry = _FourByFourRect.FourByFourRectGeometry;
exports.MergedBufferGeometry = _MergedBufferGeometry.MergedBufferGeometry;
exports.CSSBorderGeometry = _CSSBorderGeometry.CSSBorderGeometry;

// Player

exports.Overlay = _Overlay2.default;
exports.Player = _Player2.default;
